File: scratch/fetch_all.py
import os
import json
import urllib.request
import urllib.error
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime

# 설정
target_file_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'lotto.json')
latest_round = 1232
concurrency_limit = 25

def fetch_round(round_num):
    url = f"https://www.dhlottery.co.kr/common.do?method=getLottoNumber&drwNo={round_num}"
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req, timeout=10) as response:
            data = response.read().decode('utf-8')
            res = json.loads(data)
            if res.get('returnValue') == 'success':
                numbers = [
                    res.get('drwtNo1'),
                    res.get('drwtNo2'),
                    res.get('drwtNo3'),
                    res.get('drwtNo4'),
                    res.get('drwtNo5'),
                    res.get('drwtNo6')
                ]
                # 정수형 정렬
                numbers = sorted([int(n) for n in numbers])
                return {
                    "round": int(res.get('drwNo')),
                    "date": res.get('drwNoDate'),
                    "numbers": numbers,
                    "bonus": int(res.get('bonusNo')),
                    "prize": int(res.get('firstWinamnt', 0)),
                    "winners": int(res.get('firstPrzewinnerCo', 0)),
                    "sales": int(res.get('totSellamnt', 0))
                }
    except Exception as e:
        print(f"[오류] {round_num}회차 가져오기 실패: {e}")
    return None

def main():
    print(f"1회부터 {latest_round}회까지 전체 로또 데이터를 동행복권 API에서 가져오는 중...")
    results = []
    
    # 멀티스레딩으로 빠른 속도로 수집
    with ThreadPoolExecutor(max_workers=concurrency_limit) as executor:
        futures = {executor.submit(fetch_round, r): r for r in range(1, latest_round + 1)}
        for future in as_completed(futures):
            round_num = futures[future]
            res = future.result()
            if res:
                results.append(res)
                if len(results) % 100 == 0:
                    print(f"진행 완료: {len(results)}개 회차 수집 완료...")
            else:
                # 간단한 재시도 1회
                print(f"{round_num}회차 재시도 중...")
                res_retry = fetch_round(round_num)
                if res_retry:
                    results.append(res_retry)
                
    # 내림차순 정렬 (최신 회차가 맨 위로)
    results.sort(key=lambda x: x['round'], reverse=True)
    
    lotto_data = {
        "_schema": {
            "round": "회차 번호 (number)",
            "date": "추첨 날짜 ISO 8601 (string: YYYY-MM-DD)",
            "numbers": "당첨 번호 6개 오름차순 (number[])",
            "bonus": "보너스 번호 (number)",
            "prize": "1등 당첨금 (number)",
            "winners": "1등 당첨자수 (number)",
            "sales": "총 판매액 (number)"
        },
        "_source": "동행복권 (https://www.dhlottery.co.kr)",
        "_last_updated": datetime.now().strftime("%Y-%m-%d"),
        "data": results
    }
    
    os.makedirs(os.path.dirname(target_file_path), exist_ok=True)
    with open(target_file_path, 'w', encoding='utf-8') as f:
        json.dump(lotto_data, f, ensure_ascii=False, indent=2)
        
    print(f"성공! 총 {len(results)}개의 회차 정보가 {target_file_path}에 저장되었습니다.")

if __name__ == '__main__':
    main()

File: scratch/fetch-all.js
const fs = require('fs');
const https = require('https');
const path = require('path');

const targetFilePath = path.join(__dirname, '..', 'data', 'lotto.json');
const latestRound = 1232;
const concurrencyLimit = 30; // 동시 요청 수 제한

// API 호출 헬퍼
function fetchRound(round) {
  return new Promise((resolve, reject) => {
    const url = `https://www.dhlottery.co.kr/common.do?method=getLottoNumber&drwNo=${round}`;
    https.get(url, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          if (json.returnValue === 'success') {
            resolve({
              round: json.drwNo,
              date: json.drwNoDate,
              numbers: [
                json.drwtNo1,
                json.drwtNo2,
                json.drwtNo3,
                json.drwtNo4,
                json.drwtNo5,
                json.drwtNo6
              ].map(Number).sort((a, b) => a - b),
              bonus: Number(json.bonusNo),
              prize: Number(json.firstWinamnt || 0),
              winners: Number(json.firstPrzewinnerCo || 0),
              sales: Number(json.totSellamnt || 0)
            });
          } else {
            reject(new Error(`Fail to fetch round ${round}`));
          }
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', (err) => {
      reject(err);
    });
  });
}

// 청크로 나누어 병렬 처리
async function fetchAll() {
  console.log(`1회부터 ${latestRound}회까지 로또 데이터 수집을 시작합니다...`);
  const results = [];
  const rounds = Array.from({ length: latestRound }, (_, i) => i + 1);

  for (let i = 0; i < rounds.length; i += concurrencyLimit) {
    const chunk = rounds.slice(i, i + concurrencyLimit);
    console.log(`요청 진행 중: ${chunk[0]}회 ~ ${chunk[chunk.length - 1]}회...`);
    
    const promises = chunk.map(r => 
      fetchRound(r).catch(err => {
        console.error(`[오류] ${r}회차 가져오기 실패, 재시도합니다...`, err.message);
        // 재시도 1회
        return new Promise(res => setTimeout(res, 1000))
          .then(() => fetchRound(r))
          .catch(e => {
            console.error(`[오류] ${r}회차 최종 실패:`, e.message);
            return null;
          });
      })
    );

    const chunkResults = await Promise.all(promises);
    chunkResults.forEach(res => {
      if (res) results.push(res);
    });
    
    // API 서버 과부하 방지를 위한 약간의 대기
    await new Promise(res => setTimeout(res, 200));
  }

  // 내림차순 정렬 (최신이 위로)
  results.sort((a, b) => b.round - a.round);

  const lottoData = {
    _schema: {
      round: "회차 번호 (number)",
      date: "추첨 날짜 ISO 8601 (string: YYYY-MM-DD)",
      numbers: "당첨 번호 6개 오름차순 (number[])",
      bonus: "보너스 번호 (number)",
      prize: "1등 당첨금 (number)",
      winners: "1등 당첨자수 (number)",
      sales: "총 판매액 (number)"
    },
    _source: "동행복권 (https://www.dhlottery.co.kr)",
    _last_updated: new Date().toISOString().split('T')[0],
    data: results
  };

  fs.mkdirSync(path.dirname(targetFilePath), { recursive: true });
  fs.writeFileSync(targetFilePath, JSON.stringify(lottoData, null, 2), 'utf-8');
  console.log(`수집 완료! 총 ${results.length}개의 데이터가 ${targetFilePath}에 저장되었습니다.`);
}

fetchAll().catch(console.error);
